Feature/kas 2702 versies tab doc view

TaneeG opened the pull request:
Jira: https://kanselarij.atlassian.net/browse/KAS-2702

Css bekeken met @brenner-company . Deze is nu functioneel werkende voor dit ticket. Karel gaat deze op een later moment nog wat bewerken.

Commit:
KAS-2705 Added docContainer variable

Commit:
Merge branch 'feature/KAS-2834_tabs_voor_docview' into feature/KAS-2705_Versies_tab_doc_view

# Conflicts:
#	app/components/documents/document-preview-sidebar.hbs

Commit:
KAS-2834 added version cards

Commit:
KAS-2834 added version cards

Commit:
Merge branch 'feature/KAS-2834_tabs_voor_docview' into feature/KAS-2705_Versies_tab_doc_view

Commit:
KAS-2834 Added open version, check latest version , check selected version

Commit:
Merge branch 'development' into feature/KAS-2705_Versies_tab_doc_view

# Conflicts:
#	app/components/documents/document-preview-sidebar.hbs

Commit:
KAS-2705 delete version

Commit:
KAS-2705 new version modal

Commit:
Merge branch 'development' into feature/KAS-2705_Versies_tab_doc_view

# Conflicts:
#	translations/nl-be.json

Commit:
Merge branch 'development' into feature/KAS-2705_Versies_tab_doc_view

MikiDi committed:
Merge branch 'development' into feature/KAS-2705_Versies_tab_doc_view

MikiDi commented:
Ivm css: zie laatste toevoeging aan de story-beschrijving (link naar prototype)

erikap committed:
Merge branch 'development' into feature/KAS-2705_Versies_tab_doc_view

erikap committed:
Merge branch 'development' into feature/KAS-2705_Versies_tab_doc_view

erikap commented on app/components/documents/document-preview-sidebar.js:
```diff
@@ -23,13 +39,88 @@ export default class DocumentsDocumentPreviewSidebar extends Component {
+    if (isBlank(accessLevel)){
+      accessLevel = await this.store.findRecordByUri('access-level', CONSTANTS.ACCESS_LEVELS.INTERN_REGERING);
+    }
+    const now = moment().utc()
```
`const now = new Date()`

erikap commented on app/components/documents/document-preview-sidebar.js:
```diff
@@ -23,13 +39,88 @@ export default class DocumentsDocumentPreviewSidebar extends Component {
   *loadData() {
-    const docContainer = yield this.args.piece.documentContainer;
-    this.documentType = yield docContainer.type;
+    this.docContainer = yield this.args.piece.documentContainer;
+    this.loadVersionHistory.perform();
+    this.documentType = yield this.docContainer.type;
```
Hier wordt al behoorlijk wat data geladen ongeacht de tab die de gebruiker aan het bekijken is. Als de details tab bijvoorbeeld geopend is, zal toch ook al de hele versie historiek gedownload worden.

Het is waarschijnlijk beter om met aparte componenten te werken voor de inhoud van iedere tab. Op die manier gaat de `document-preview-sidebar` component ook eenvoudiger worden en enkel de logica van de tab-selectie bevatten. Nieuwe components kunnen misschien best samengezet worden in een subfolder `documents/document-preview/details`, `documents/document-preview/versions`, ...

TaneeG replied:
Ik zal het details tab component doen in https://github.com/kanselarij-vlaanderen/kaleidos-frontend/pull/1058

erikap commented on app/components/documents/document-preview-sidebar.js:
```diff
@@ -23,13 +39,88 @@ export default class DocumentsDocumentPreviewSidebar extends Component {
 
+  @task
+  *loadVersionHistory() {
+    this.versionPieces = yield this.docContainer.reverseSortedPieces;
+    this.lastPiece = yield this.docContainer.lastPiece;
```
Probeer zo veel mogelijk weg te blijven van het gebruik van de async computed properties op de ember-data models. Het verleden heeft al geillustreerd dat dit leidt tot ongecontroleerde data stromen en network requests. `this.docContainer.reverseSortedPieces` is immers heel snel gebruikt zonder de impact in te schatten van de hoeveelheid network requests dat hierdoor mogelijk getriggerd worden. Bovendien zijn heel veel van die computed properties ook weer afhankelijkheden van andere computed properties waardoor ook deze weer getriggerd worden enz.

Je kan best kijken hoe de history opgebouwd wordt in `Documents::DocumentCard` en een gelijkaardige werkwijze toepassen.

erikap commented on app/components/documents/document-preview-sidebar.js:
```diff
@@ -23,13 +39,88 @@ export default class DocumentsDocumentPreviewSidebar extends Component {
+
+    this.isOpenUploadVersionModal = false;
+    this.loadVersionHistory.perform();
+    this.router.transitionTo('document', newPiece.id)
```
Het is niet de bedoeling om steeds een nieuwe tab te openen wanneer de gebruiker een andere versie wil bekijken maar om de geselecteerde versie in de huidige preview up te daten.

erikap commented on app/styles/au-kaleidos-css/_auk-additions.scss:
```diff
@@ -287,3 +287,14 @@ p {
 }
+
+
+/* Document-viewer
```
Als er toch nog naar de styling gekeken moet worden, zou ik vermijden om zelf al custom styling toe te voegen

TaneeG replied:
Deze styling was voor de functionaliteit van het ticket te behalen

erikap reviewed (changes requested):
Nog niet naar alle code in depth gekeken, maar de eerste bevindingen waar zeker al mee begonnen kan worden:

- probeer zinvolle en consistente namen te gebruiken voor methods, variabelen en components zodat het duidelijk is wat ze betekenen (`@openVerify` heeft bijvoorbeeld weinig betekenis als argument van een component; `piece`, `version`, `versionPiece`, `document`, ... worden allemaal door elkaar gebruikt)

- kijk goed welke componenten er al zijn en als je dan toch een nieuwe component maakt voor een bestaande functionaliteit (vb. het uploaden van een versie) bekijk dan of je die nieuwe component ook kan introduceren in de bestaande implementatie zodat we geen dubbele codebase moeten onderhouden. 
- kan je de indentatie van de Handlebars files eens nakijken in je editor, want jouw editor lijkt de enige te zijn die de component arguments anders (met meer spaties) indenteert. Ik zou ook consequenter zijn in de line breaks: ofwel alle argumenten van een component op 1 lijn ofwel 1 argument per lijn, maar nu is het een mix van vanalles.

Commit:
Merge branch 'development' into feature/KAS-2702_Versies_tab_doc_view

Commit:
KAS-2702 Fix identations with prettier

Commit:
KAS-2702 Rename variables & methods, changed new date , Split loadData per tab

Commit:
KAS-2702 move new components to subfolder

Commit:
KAS-2702 versions tab component

Commit:
KAS-2702 Re use new versions saving

Commit:
KAS-2702 Change way versions is loaded

Commit:
KAS-2702fix lint

erikap committed:
Merge branch 'feature/KAS-2705_Versies_tab_doc_view' of github.com:kanselarij-vlaanderen/kaleidos-frontend into feature/KAS-2705_Versies_tab_doc_view

erikap commented on app/components/documents/document-card.js:
```diff
@@ -113,57 +123,45 @@ export default class DocumentsDocumentCardComponent extends Component {
   async openUploadModal() {
-    if (this.args.onOpenUploadModal) {
-      await this.args.onOpenUploadModal();
-    }
```
Waarom zijn deze regels verwijderd? Die lijken mij nog steeds gebruikt te worden. Bvb. via de `Documents::DocumentCardList` in de `agenda.agendaitems.agendaitem.documents` route.

erikap commented on app/components/documents/document-preview/document-preview-sidebar.hbs:
```diff
@@ -0,0 +1,105 @@
+    {{! VERSIONS }}
+  {{else if (eq this.activeTab "versions")}}
+    <Utils::LoadableContent @isLoading={{this.loadVersionsData.isRunning}}>
+      <Documents::DocumentPreview::Versions::VersionsTab
```
Ik zou persoonlijk geen te diepe nesting maken van componenten en het gewoon bij `Documents::DocumentPreview::VersionsTab` houden (en dan later ook `Documents::DocumentPreview::DetailsTab` en `Documents::DocumentPreview::SignaturesTab`)

erikap commented on app/components/documents/document-preview/document-preview-sidebar.hbs:
```diff
@@ -0,0 +1,105 @@
+      <Documents::DocumentPreview::Versions::VersionsTab
+        @piece={{@piece}}
+        @lastPiece={{this.lastPiece}}
+        @versions={{this.versions}}
```
Als ik logisch nadenk over die versions-tab component zou ik als argument meegeven:
- `@documentContainer`: de container met alle versies
- `@currentSelectedPiece`: het geselecteerde document dat momenteel in de preview getoond wordt
- `@onChangeSelectedPiece`: action-handler wanneer een nieuwe versie aangeklikt wordt in de versions-tab. In die action handler kan dan de document preview geupdate worden

Alle logica rond volgorde van de pieces, welk stuk het laatste piece is etc. hoor in mijn ogen thuis in die versies-tab

erikap commented on app/components/documents/document-preview/document-preview-sidebar.js:
```diff
@@ -0,0 +1,138 @@
+    const pieces = yield this.documentContainer.hasMany('pieces').reload();
+    const sortedPieces = A(sortPieces(pieces.toArray()).reverse());
+    this.versions = sortedPieces.slice(0).reverse();
+    this.lastPiece = sortedPieces.lastObject;
```
Zie ook vorige comment. Dit hoort voor mij logischerwijs in de versions-tab component thuis.

erikap commented on app/components/documents/document-preview/document-preview-sidebar.js:
```diff
@@ -0,0 +1,138 @@
+  cancelDeletePiece() {
+    this.selectedToDelete = null;
+    this.isDeletingPiece = false;
+  }
```
Door deze PR te reviewen ben ik tot het besef gekomen dat het toevoegen van een nieuwe documentversie niet zo triviaal is. De acties die moeten gebeuren nadat de versie opgeladen is, zijn afhankelijk van de context (route). Bijvoorbeeld bij een agendapunt wordt die nieuwe versie ook gelinkt aan het agendapunt, bij de beslissingsfiche aan de beslissingsactivieit, .... 
Als we die modal in een nieuwe route in een nieuwe tab openen, dan hebben we die context niet en gaat het document niet correct aan andere resources gelinkt worden. En het document openen in een nieuwe route is momenteel de enige use case die we hebben.

Ik denk daarom dat het beter is de scope van de story in te perken tot:
- het weergeven van de lijst van versies
- het selecteren van een versie om te bekijken

en het uploaden/verwijderen van een nieuwe versie weg te laten op dit moment.

Het uploaden/verwijderen moeten we dan case per case gaan uitwerken op het moment dat we gaan ondersteunen dat de document-preview op bepaalde plaatsen effectief als een in-app modal geopend wordt.

erikap commented on app/components/documents/document-card.js:
```diff
@@ -113,57 +123,45 @@ export default class DocumentsDocumentCardComponent extends Component {
+        documentContainer: this.documentContainer,
+      });
+      yield newPiece.save();
+
```
Ik ben niet helemaal mee waarom deze wijziging hier gebeurt. Ik vermoed dat het extraheren van het document-upload gedeelte er iets mee te maken heeft?

Deze changes lijken mij breaking te zijn voor andere plaatsen waar de `Documents::DocumentCard` gebruikt wordt. Ik zie bijvoorbeeld de callback `onAddPiece` die niet meer aangeroepen wordt, of de `save()` van het nieuwe `piece`-record dat voordien niet in deze component gebeurde en nu wel.

Gegeven de vorige comment over het beperken van de scope en de version upload weg te laten, lijkt het mij het veiligst om de changes op de `Documents::DocumentCard` component volledig te rollbacken want die component is key voor de huidige applicatie in productie.

erikap commented on app/pods/document/controller.js:
```diff
@@ -14,4 +14,9 @@ export default class DocumentController extends Controller {
+
+  @action
+  openNewPiece(piece) {
+    this.model = piece;
```
Het model updaten in de controller is een anti-pattern dat je moet vermijden. Indien het model aangepast moet worden, moet die via de `model()`-hook.

Nu, voor dit geval lijkt het mij eerder state van de `document-preview-modal` component te zijn. Logisch gezien krijgt die component voor mij het document uit de model hook mee als argument (`@piece` nu) en wordt dit gebruikt om zijn state `this.currentSelectedPiece` te initializeren. Als er in de versies tab een nieuwe versie gekozen wordt, past de `document-preview-modal` zijn state aan, maar die change lijkt mij niet tot hier te moeten door bubblen.

erikap commented on app/components/documents/document-preview/versions/version-card.js:
```diff
@@ -0,0 +1,46 @@
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+
+export default class  DocumentsDocumentPreviewVersionsVersionCardComponent extends Component {
```
Ik zou persoonlijk geen te diepe nesting maken van componenten en het gewoon bij `Documents::DocumentPreview::VersionCard` houden

erikap commented on app/components/documents/document-preview/versions/version-card.hbs:
```diff
@@ -0,0 +1,54 @@
+<div class="auk-u-mb-4">
```
Het switchen tussen getoonde document-versie in de preview door een version-card aan te klikken (of beter een button erop) lijkt nog niet te werken?

erikap committed:
Update styling of version-card in document preview

erikap reviewed (changes requested):
Morgen te bekijken in de standup hoe we verder gaan met deze story (zie verdere comments rond version upload)

Commit:
Revert "KAS-2702 Re use new versions saving"

This reverts commit d0241e3fc0dabde8aa0c15ebad2bedf560cece07.

Commit:
KAS-2702 await loadversion

Commit:
Merge remote-tracking branch 'origin/feature/KAS-2705_Versies_tab_doc_view' into feature/KAS-2702_Versies_tab_doc_view

Commit:
KAS-2702 Add lost code version history task

Commit:
KAS-2702 undo deeper nesting

Commit:
KAS-2702 Remove delete & add version

Commit:
KAS-2702 cleanup with prettier

erikap committed:
Merge branch 'development' into feature/KAS-2705_Versies_tab_doc_view

erikap committed:
Track current selected document version as property in document-preview-modal

erikap committed:
Remove unused service injections from document-preview-sidebar

erikap committed:
Fix contextual components for toolbar-group and toolbar-item

erikap committed:
Refactor toolbar to contextual components

erikap committed:
Remove duplicate array reverse

erikap committed:
Refactor sorting to get-function

erikap committed:
Add isHighlighted argument on version-card

Using isHighligthed to highlight current selected version. In line
with document-card component.

erikap committed:
Remove unused service injection from version-card
